File: example/src/main/java/com/xwray/groupie/example/MainActivity.java
package com.xwray.groupie.example;

import android.content.Intent;
import android.content.SharedPreferences;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.os.Handler;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.text.TextUtils;
import android.view.View;
import android.widget.Toast;

import com.xwray.groupie.example.R;
import com.xwray.groupie.ExpandableGroup;
import com.xwray.groupie.GroupAdapter;
import com.xwray.groupie.Item;
import com.xwray.groupie.OnItemClickListener;
import com.xwray.groupie.Section;
import com.xwray.groupie.TouchCallback;
import com.xwray.groupie.UpdatingGroup;
import com.xwray.groupie.example.databinding.ActivityMainBinding;
import com.xwray.groupie.example.decoration.CarouselItemDecoration;
import com.xwray.groupie.example.decoration.DebugItemDecoration;
import com.xwray.groupie.example.decoration.HeaderItemDecoration;
import com.xwray.groupie.example.decoration.InsetItemDecoration;
import com.xwray.groupie.example.decoration.SwipeTouchCallback;
import com.xwray.groupie.example.item.CardItem;
import com.xwray.groupie.example.item.CarouselCardItem;
import com.xwray.groupie.example.item.CarouselItem;
import com.xwray.groupie.example.item.ColumnItem;
import com.xwray.groupie.example.item.FullBleedCardItem;
import com.xwray.groupie.example.item.HeaderItem;
import com.xwray.groupie.example.item.HeartCardItem;
import com.xwray.groupie.example.item.SmallCardItem;
import com.xwray.groupie.example.item.SwipeToDeleteItem;
import com.xwray.groupie.example.item.UpdatableItem;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    public static final String INSET_TYPE_KEY = "inset_type";
    public static final String FULL_BLEED = "full_bleed";
    public static final String INSET = "inset";

    private ActivityMainBinding binding;
    private StickyGroupAdapter groupAdapter;
    private GridLayoutManager layoutManager;
    private Prefs prefs;

    private int gray;
    private int betweenPadding;
    private int[] rainbow200;
    private int[] rainbow500;

    private Section infiniteLoadingSection;
    private Section swipeSection;

    // Normally there's no need to hold onto a reference to this list, but for demonstration
    // purposes, we'll shuffle this list and post an update periodically
    private ArrayList<UpdatableItem> updatableItems;

    // Hold a reference to the updating group, so we can, well, update it
    private UpdatingGroup updatingGroup;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        prefs = Prefs.get(this);

        gray = ContextCompat.getColor(this, R.color.background);
        betweenPadding = getResources().getDimensionPixelSize(R.dimen.padding_small);
        rainbow200 = getResources().getIntArray(R.array.rainbow_200);
        rainbow500 = getResources().getIntArray(R.array.rainbow_500);

        groupAdapter = new StickyGroupAdapter(R.layout.item_header, R.layout.item_expandable_header);
        groupAdapter.setOnItemClickListener(onItemClickListener);
        groupAdapter.setSpanCount(12);
        populateAdapter();
        layoutManager = new GridLayoutManager(this, groupAdapter.getSpanCount());
        layoutManager.setSpanSizeLookup(groupAdapter.getSpanSizeLookup());

        final RecyclerView recyclerView = binding.recyclerView;
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.addItemDecoration(new HeaderItemDecoration(gray, betweenPadding));
        recyclerView.addItemDecoration(new InsetItemDecoration(gray, betweenPadding));
        recyclerView.addItemDecoration(new DebugItemDecoration(this));
        recyclerView.setAdapter(groupAdapter);
        recyclerView.addOnScrollListener(new InfiniteScrollListener(layoutManager) {
            @Override public void onLoadMore(int currentPage) {
                int color = rainbow200[currentPage % rainbow200.length];
                for (int i = 0; i < 5; i++) {
                    infiniteLoadingSection.add(new CardItem(color));
                }
            }
        });

        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(touchCallback);
        itemTouchHelper.attachToRecyclerView(recyclerView);

        binding.fab.setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View view) {
                startActivity(new Intent(MainActivity.this, SettingsActivity.class));
            }
        });

        prefs.registerListener(onSharedPrefChangeListener);
    }

    private void populateAdapter() {

        // Full bleed item
        Section fullBleedItemSection = new Section(new HeaderItem(R.string.full_bleed_item));
        fullBleedItemSection.add(new FullBleedCardItem(R.color.purple_200));
        groupAdapter.add(fullBleedItemSection);

        // Update in place group
        Section updatingSection = new Section();
        View.OnClickListener onShuffleClicked = new View.OnClickListener() {
            @Override public void onClick(View view) {
                List<UpdatableItem> shuffled = new ArrayList<>(updatableItems);
                Collections.shuffle(shuffled);
                updatingGroup.update(shuffled);

                // You can also do this by forcing a change with payload
                binding.recyclerView.post(new Runnable() {
                    @Override public void run() {
                        binding.recyclerView.invalidateItemDecorations();
                    }
                });
            }
        };
        HeaderItem updatingHeader = new HeaderItem(
                R.string.updating_group,
                R.string.updating_group_subtitle,
                R.drawable.shuffle,
                onShuffleClicked);
        updatingSection.setHeader(updatingHeader);
        updatingGroup = new UpdatingGroup();
        updatableItems = new ArrayList<>();
        for (int i = 1; i <= 12; i++) {
            updatableItems.add(new UpdatableItem(rainbow200[i], i));
        }
        updatingGroup.update(updatableItems);
        updatingSection.add(updatingGroup);
        groupAdapter.add(updatingSection);

        // Expandable group
        ExpandableHeaderItem expandableHeaderItem = new ExpandableHeaderItem(R.string.expanding_group, R.string.expanding_group_subtitle);
        ExpandableGroup expandableGroup = new ExpandableGroup(expandableHeaderItem);
        for (int i = 0; i < 2; i++) {
            expandableGroup.add(new CardItem(rainbow200[1]));
        }
        groupAdapter.add(expandableGroup);

        // Columns
        Section columnSection = new Section(new HeaderItem(R.string.vertical_columns));
        ColumnGroup columnGroup = makeColumnGroup();
        columnSection.add(columnGroup);
        groupAdapter.add(columnSection);

        // Group showing even spacing with multiple columns
        Section multipleColumnsSection = new Section(new HeaderItem(R.string.multiple_columns));
        for (int i = 0; i < 12; i++) {
            multipleColumnsSection.add(new SmallCardItem(rainbow200[5]));
        }
        groupAdapter.add(multipleColumnsSection);

        // Swipe to delete (with add button in header)
        swipeSection = new Section(new HeaderItem(R.string.swipe_to_delete));
        for (int i = 0; i < 3; i++) {
            swipeSection.add(new SwipeToDeleteItem(rainbow200[6]));
        }
        groupAdapter.add(swipeSection);

        // Horizontal carousel
        Section carouselSection = new Section(new HeaderItem(R.string.carousel, R.string.carousel_subtitle));
        CarouselItem carouselItem = makeCarouselItem();
        carouselSection.add(carouselItem);
        groupAdapter.add(carouselSection);

        // Update with payload
        Section updateWithPayloadSection = new Section(new HeaderItem(R.string.update_with_payload, R.string.update_with_payload_subtitle));
        for (int i = 0; i < rainbow500.length; i++) {
            updateWithPayloadSection.add(new HeartCardItem(rainbow200[i], i, onFavoriteListener));

        }
        groupAdapter.add(updateWithPayloadSection);

        // Infinite loading section
        infiniteLoadingSection = new Section(new HeaderItem(R.string.infinite_loading));
        groupAdapter.add(infiniteLoadingSection);
    }

    private ColumnGroup makeColumnGroup() {
        List<ColumnItem> columnItems = new ArrayList<>();
        for (int i = 1; i <= 5; i++) {
            // First five items are red -- they'll end up in a vertical column
            columnItems.add(new ColumnItem(rainbow200[0], i));
        }
        for (int i = 6; i <= 10; i++) {
            // Next five items are pink
            columnItems.add(new ColumnItem(rainbow200[1], i));
        }
        return new ColumnGroup(columnItems);
    }

    private CarouselItem makeCarouselItem() {
        CarouselItemDecoration carouselDecoration = new CarouselItemDecoration(gray, betweenPadding);
        GroupAdapter carouselAdapter = new GroupAdapter();
        for (int i = 0; i < 30; i++) {
            carouselAdapter.add(new CarouselCardItem(rainbow200[7]));
        }
        CarouselItem carouselItem = new CarouselItem(carouselDecoration);
        carouselItem.setAdapter(carouselAdapter);
        return carouselItem;
    }

    private OnItemClickListener onItemClickListener = new OnItemClickListener() {
        @Override
        public void onItemClick(Item item, View view) {
            if (item instanceof CardItem) {
                CardItem cardItem = (CardItem) item;
                if (!TextUtils.isEmpty(cardItem.getText())) {
                    Toast.makeText(MainActivity.this, cardItem.getText(), Toast.LENGTH_SHORT).show();
                }
            }
        }
    };

    @Override protected void onDestroy() {
        prefs.unregisterListener(onSharedPrefChangeListener);
        super.onDestroy();
    }

    private TouchCallback touchCallback = new SwipeTouchCallback(gray) {
        @Override public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
            return false;
        }

        @Override public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
            Item item = groupAdapter.getItem(viewHolder.getAdapterPosition());
            // Change notification to the adapter happens automatically when the section is
            // changed.
            swipeSection.remove(item);
        }
    };

    private SharedPreferences.OnSharedPreferenceChangeListener onSharedPrefChangeListener =
            new SharedPreferences.OnSharedPreferenceChangeListener() {
                @Override public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String s) {
                    // This is pretty evil, try not to do this
                    groupAdapter.notifyDataSetChanged();
                }
            };

    private Handler handler = new Handler();
    private HeartCardItem.OnFavoriteListener onFavoriteListener = new HeartCardItem.OnFavoriteListener() {
        @Override
        public void onFavorite(final HeartCardItem item, final boolean favorite) {
            // Pretend to make a network request
            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    // Network request was successful!
                    item.setFavorite(favorite);
                    item.notifyChanged(HeartCardItem.FAVORITE);
                }
            }, 1000);
        }
    };

}

File: example/src/main/java/com/xwray/groupie/example/MainActivity.kt
package com.xwray.groupie.example

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.os.Handler
import android.text.TextUtils
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.recyclerview.selection.SelectionPredicates
import androidx.recyclerview.selection.SelectionTracker
import androidx.recyclerview.selection.StableIdKeyProvider
import androidx.recyclerview.selection.StorageStrategy
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.xwray.groupie.*
import com.xwray.groupie.example.core.InfiniteScrollListener
import com.xwray.groupie.example.core.Prefs
import com.xwray.groupie.example.core.SettingsActivity
import com.xwray.groupie.example.core.decoration.CarouselItemDecoration
import com.xwray.groupie.example.core.decoration.DebugItemDecoration
import com.xwray.groupie.example.core.decoration.SwipeTouchCallback
import com.xwray.groupie.example.item.*
import com.xwray.groupie.groupiex.plusAssign
import kotlinx.android.synthetic.main.activity_main.*

const val INSET_TYPE_KEY = "inset_type"
const val INSET = "inset"

class MainActivity : AppCompatActivity() {

    private val groupAdapter = GroupAdapter<GroupieViewHolder>() //TODO get rid of this parameter
    private lateinit var groupLayoutManager: GridLayoutManager
    private val prefs by lazy { Prefs.get(this) }
    private val handler = Handler()

    private val gray: Int by lazy { ContextCompat.getColor(this, R.color.background) }
    private val betweenPadding: Int by lazy { resources.getDimensionPixelSize(R.dimen.padding_small) }
    private val rainbow200: IntArray by lazy { resources.getIntArray(R.array.rainbow_200) }
    private val rainbow500: IntArray by lazy { resources.getIntArray(R.array.rainbow_500) }

    private val infiniteLoadingSection = Section(HeaderItem(R.string.infinite_loading))
    private val swipeSection = Section(HeaderItem(R.string.swipe_to_delete))

    private var tracker: SelectionTracker<Long>? = null

    // Normally there's no need to hold onto a reference to this list, but for demonstration
    // purposes, we'll shuffle this list and post an update periodically
    private val updatableItems: ArrayList<UpdatableItem> by lazy {
        ArrayList<UpdatableItem>().apply {
            for (i in 1..12) {
                add(UpdatableItem(rainbow200[i], i))
            }
        }
    }

    // Hold a reference to the updating group, so we can, well, update it
    private var updatingGroup = Section()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        groupAdapter.apply {
            setOnItemClickListener(onItemClickListener)
            setOnItemLongClickListener(onItemLongClickListener)
            spanCount = 12
        }

        recreateAdapter()

        groupLayoutManager = GridLayoutManager(this, groupAdapter.spanCount).apply {
            spanSizeLookup = groupAdapter.spanSizeLookup
        }

        recyclerView.apply {
            layoutManager = groupLayoutManager
            addItemDecoration(HeaderItemDecoration(gray, betweenPadding))
            addItemDecoration(InsetItemDecoration(gray, betweenPadding))
            addItemDecoration(DebugItemDecoration(context))
            adapter = groupAdapter
            addOnScrollListener(object : InfiniteScrollListener(groupLayoutManager) {
                override fun onLoadMore(currentPage: Int) {
                    val color = rainbow200[currentPage % rainbow200.size]
                    for (i in 0..4) {
                        infiniteLoadingSection.add(CardItem(color))
                    }
                }
            })
        }

        tracker = SelectionTracker.Builder<Long>(
                "groupie-selection",
                recyclerView,
                StableIdKeyProvider(recyclerView),
                GroupieLookup(recyclerView),
                StorageStrategy.createLongStorage()
        ).withSelectionPredicate(SelectionPredicates.createSelectAnything())
                .build()
        groupAdapter.setSelectionTracker(tracker)

        ItemTouchHelper(touchCallback).attachToRecyclerView(recyclerView)

        fab.setOnClickListener { startActivity(Intent(this@MainActivity, SettingsActivity::class.java)) }

        prefs.registerListener(onSharedPrefChangeListener)

    }

    private fun recreateAdapter() {
        groupAdapter.clear()
        groupAdapter.setHasStableIds(true)

        if (prefs.useAsync) {
            populateAdapterAsync()
        } else {
            populateAdapter()
        }
    }

    private fun populateAdapter() {

        // Full bleed item
        groupAdapter += Section(HeaderItem(R.string.full_bleed_item)).apply {
            add(FullBleedCardItem(R.color.purple_200))
        }

        // Update in place group
        groupAdapter += Section().apply {
            val updatingHeader = HeaderItem(
                    R.string.updating_group,
                    R.string.updating_group_subtitle,
                    R.drawable.shuffle,
                    onShuffleClicked)
            setHeader(updatingHeader)

            updatingGroup.update(updatableItems)
            add(updatingGroup)
        }

        // Expandable group
        val expandableHeaderItem = ExpandableHeaderItem(R.string.expanding_group, R.string.expanding_group_subtitle)
        groupAdapter += ExpandableGroup(expandableHeaderItem).apply {
            for (i in 0..1) {
                add(SelectionItem(rainbow200[1]))
            }
        }

        // Reordering a Section of Expandable Groups
        val section = Section(HeaderItem(R.string.reorderable_section))
        val swappableExpandableGroup = mutableListOf<ExpandableGroup>()
        var colorIndex = 0
        for (i in 0..2) {
            val header = ExpandableHeaderItem(R.string.reorderable_item_title, R.string.reorderable_item_subtitle)
            val group = ExpandableGroup(header).apply {
                val numChildren= i * 2 // groups will continue to grow by 2
                for (j in 0..numChildren) {
                    add(CardItem(rainbow200[colorIndex]))
                    if (colorIndex + 1 >= rainbow200.size) {
                        colorIndex = 0
                    } else {
                        colorIndex += 1
                    }
                }
            }
            header.clickListener = {
                swappableExpandableGroup.remove(group)
                swappableExpandableGroup.add(group)

                section.update(swappableExpandableGroup)
            }
            swappableExpandableGroup.add(group)
        }
        section.addAll(swappableExpandableGroup)
        groupAdapter += section

        // Columns
        groupAdapter += Section(HeaderItem(R.string.vertical_columns)).apply {
            add(makeColumnGroup())
        }

        // Group showing even spacing with multiple columns
        groupAdapter += Section(HeaderItem(R.string.multiple_columns)).apply {
            for (i in 0..11) {
                add(SmallCardItem(rainbow200[5]))
            }
        }

        // Swipe to delete (with add button in header)
        for (i in 0..2) {
            swipeSection += SwipeToDeleteItem(rainbow200[6])
        }
        groupAdapter += swipeSection

        // Horizontal carousel
        groupAdapter += Section(HeaderItem(R.string.carousel, R.string.carousel_subtitle)).apply {
            add(makeCarouselItem())
        }

        // Update with payload
        groupAdapter += Section(HeaderItem(R.string.update_with_payload, R.string.update_with_payload_subtitle)).apply {
            rainbow500.indices.forEach { i ->
                add(HeartCardItem(rainbow200[i], i.toLong()) { item, favorite ->
                    // Pretend to make a network request
                    handler.postDelayed({
                        // Network request was successful!
                        item.setFavorite(favorite)
                        item.notifyChanged(FAVORITE)
                    }, 1000)
                })
            }
        }

        // Infinite loading section
        groupAdapter += infiniteLoadingSection
    }

    private fun populateAdapterAsync() {

        val allGroups = mutableListOf<Group>()

        // Full bleed item
        allGroups += Section(HeaderItem(R.string.full_bleed_item)).apply {
            add(FullBleedCardItem(R.color.purple_200))
        }

        // Update in place group
        allGroups += Section().apply {
            val updatingHeader = HeaderItem(
                R.string.updating_group,
                R.string.updating_group_subtitle,
                R.drawable.shuffle,
                onShuffleClicked)
            setHeader(updatingHeader)

            updatingGroup.update(updatableItems)
            add(updatingGroup)
        }

        // Expandable group
        val expandableHeaderItem = ExpandableHeaderItem(R.string.expanding_group, R.string.expanding_group_subtitle)
        allGroups += ExpandableGroup(expandableHeaderItem).apply {
            for (i in 0..1) {
                add(CardItem(rainbow200[1]))
            }
        }

        // Reordering a Section of Expandable Groups
        val section = Section(HeaderItem(R.string.reorderable_section))
        val swappableExpandableGroup = mutableListOf<ExpandableGroup>()
        var colorIndex = 0
        for (i in 0..2) {
            val header = ExpandableHeaderItem(R.string.reorderable_item_title, R.string.reorderable_item_subtitle)
            val group = ExpandableGroup(header).apply {
                val numChildren= i * 2 // groups will continue to grow by 2
                for (j in 0..numChildren) {
                    add(CardItem(rainbow200[colorIndex]))
                    if (colorIndex + 1 >= rainbow200.size) {
                        colorIndex = 0
                    } else {
                        colorIndex += 1
                    }
                }
            }
            header.clickListener = {
                swappableExpandableGroup.remove(group)
                swappableExpandableGroup.add(group)

                section.update(swappableExpandableGroup)
            }
            swappableExpandableGroup.add(group)
        }
        section.addAll(swappableExpandableGroup)
        allGroups += section

        // Columns
        allGroups += Section(HeaderItem(R.string.vertical_columns)).apply {
            add(makeColumnGroup())
        }

        // Group showing even spacing with multiple columns
        allGroups += Section(HeaderItem(R.string.multiple_columns)).apply {
            for (i in 0..11) {
                add(SmallCardItem(rainbow200[5]))
            }
        }

        // Swipe to delete (with add button in header)
        for (i in 0..2) {
            swipeSection += SwipeToDeleteItem(rainbow200[6])
        }
        allGroups += swipeSection

        // Horizontal carousel
        allGroups += Section(HeaderItem(R.string.carousel, R.string.carousel_subtitle)).apply {
            add(makeCarouselItem())
        }

        // Update with payload
        allGroups += Section(HeaderItem(R.string.update_with_payload, R.string.update_with_payload_subtitle)).apply {
            rainbow500.indices.forEach { i ->
                add(HeartCardItem(rainbow200[i], i.toLong()) { item, favorite ->
                    // Pretend to make a network request
                    handler.postDelayed({
                        // Network request was successful!
                        item.setFavorite(favorite)
                        item.notifyChanged(FAVORITE)
                    }, 1000)
                })
            }
        }

        // Infinite loading section
        allGroups += infiniteLoadingSection

        groupAdapter.updateAsync(allGroups)
    }

    private fun makeColumnGroup(): ColumnGroup {
        val columnItems = ArrayList<ColumnItem>()
        for (i in 1..5) {
            // First five items are red -- they'll end up in a vertical column
            columnItems += ColumnItem(rainbow200[0], i)
        }
        for (i in 6..10) {
            // Next five items are pink
            columnItems += ColumnItem(rainbow200[1], i)
        }
        return ColumnGroup(columnItems)
    }

    private fun makeCarouselItem(): CarouselItem {
        val carouselDecoration = CarouselItemDecoration(gray, betweenPadding)
        val carouselAdapter = GroupAdapter<GroupieViewHolder>()
        for (i in 0..29) {
            carouselAdapter += CarouselCardItem(rainbow200[7])
        }
        return CarouselItem(carouselDecoration, carouselAdapter)
    }

    private val onItemClickListener = OnItemClickListener { item, _ ->
        if (item is CardItem && !TextUtils.isEmpty(item.text)) {
            Toast.makeText(this@MainActivity, item.text, Toast.LENGTH_SHORT).show()
        }
    }

    private val onItemLongClickListener = OnItemLongClickListener { item, _ ->
        if (item is CardItem && !item.text.isNullOrBlank()) {
            Toast.makeText(this@MainActivity, "Long clicked: " + item.text, Toast.LENGTH_SHORT).show()
            return@OnItemLongClickListener true
        }
        false
    }

    private val onShuffleClicked = View.OnClickListener {
        ArrayList(updatableItems).apply {
            shuffle()
            updatingGroup.update(this)
        }

        // You can also do this by forcing a change with payload
        recyclerView.post { recyclerView.invalidateItemDecorations() }
    }

    override fun onDestroy() {
        prefs.unregisterListener(onSharedPrefChangeListener)
        super.onDestroy()
    }

    private val touchCallback: SwipeTouchCallback by lazy {
        object : SwipeTouchCallback(gray) {
            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder,
                                target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val item = groupAdapter.getItem(viewHolder.adapterPosition)
                // Change notification to the adapter happens automatically when the section is
                // changed.
                swipeSection.remove(item)
            }
        }
    }

    private val onSharedPrefChangeListener = SharedPreferences.OnSharedPreferenceChangeListener { _, _ ->
        // This is pretty evil, try not to do this
        recreateAdapter()
    }

}
